openfl/experimental/workflow/interface/fl_spec.py

```diff
@@ -132,7 +123,110 @@ def runtime(self, runtime) -> None:
             raise TypeError(f"{runtime} is not a valid OpenFL Runtime")
         self._runtime = runtime
 
-    def _capture_instance_snapshot(self, kwargs):
+    def run(self) -> None:
+        """Starts the execution of the flow."""
+        # Submit flow to Runtime
+        if str(self._runtime) == "LocalRuntime":
+            self._run_local()
+        elif str(self._runtime) == "FederatedRuntime":
+            self._run_federated()
+        else:
+            raise Exception("Runtime not implemented")
+
+    def _run_local(self) -> None:
+        """Executes the flow using LocalRuntime."""
+        self._setup_initial_state()
+        try:
+            # Execute all Participant (Aggregator & Collaborator) tasks and
+            # retrieve the final attributes
+            # start step is the first task & invoked on aggregator through
+            # runtime.execute_task
+            final_attributes = self.runtime.execute_task(
+                self,
+                self.start,
+            )
+        except Exception as e:
+            if "cannot pickle" in str(e) or "Failed to unpickle" in str(e):
+                msg = (
+                    "\nA serialization error was encountered that could not"
+                    "\nbe handled by the ray backend."
+                    "\nTry rerunning the flow without ray as follows:\n"
+                    "\nLocalRuntime(...,backend='single_process')\n"
+                    "\n or for more information about the original error,"
+                    "\nPlease see the official Ray documentation"
+                    "\nhttps://docs.ray.io/en/releases-2.2.0/ray-core/\
+                    objects/serialization.html"
+                )
+                raise SerializationError(str(e) + msg)
+            else:
+                raise e
+        for name, attr in final_attributes:
+            setattr(self, name, attr)
+
+    def _setup_initial_state(self) -> None:
+        """
+        Sets up the flow's initial state, initializing private attributes for
+        collaborators and aggregators.
+        """
+        self._metaflow_interface = MetaflowInterface(self.__class__, self.runtime.backend)
+        self._run_id = self._metaflow_interface.create_run()
+        # Initialize aggregator private attributes
+        self.runtime.initialize_aggregator()
+        self._foreach_methods = []
+        FLSpec._reset_clones()
```

Comment:
by the way, if the runtime were a separate object from the `FLSpec`, it could also manage the clones. This would enable the `FLSpec` to become stateless - effectively a "declarative" specification of what should be executed where, but not how.

Reply:
Absolutely. This aspect will also be covered as part of analysis of refactoring runtimes https://github.com/securefederatedai/openfl/pull/1190#discussion_r1877585581